### PR TITLE
sync: resize bounded channels

### LINKED ISSUE #3730: Allow bounded mpsc channels to have resizable capacity
**Motivation**
I have a worker which pops tasks off of a work queue, which in this case is a Tokio channel. Tasks come in over the network from a single sender and the worker is able to dynamically give the sender an upper bound for the number of tasks it's willing to buffer depending on how fast it can process incoming tasks. Tokio doesn't really have a good solution for this use case. Bounded channels can only have a fixed capacity, and unbounded channels don't have a public API for how many items are currently in the channel.

**Potential solution**
I'd like to see either some sort of resizing functionality added to bounded channels, or a new type of channel with resizing functionality if making bounded channels resizable would significantly change the implementation details of bounded channels. I'm picturing it as having a `resize` method available on the `Receiver` side of the channel.

**Alternatives**
Right now my internal workaround will be to have my Tokio worker aggressively move incoming tasks from a small bounded queue to a larger dynamically sized `VecDeque`, but this is not ideal since the channel may erroneously drop tasks if the scheduler doesn't give the worker time to shift tasks out of the channel and into the deque.

### PR DESCRIPTION
**Pull request by aym-v**
Proposed solution to close #3730

## Motivation

Bounded channels can only have a fixed capacity. When a receiver wants to process more messages at a time than the originally intended capacity, it cannot resize the channel to do so.

## Solution

Implement a method that allows the user to resize the channel from the receiver handle

### CODE REVIEW

**Review by Darksonn — COMMENTED** [review]
Can't this be done in an easier way? If the method is only available for the receiver, we can just use a non-atomic integer in the receiver since there is only one receiver.

**Comment by aym-v** [comment]
@Darksonn I moved the underflow into the receiver to avoid using an atomic. This new implementation is somewhat simpler, although it can probably be improved.

File: tokio/src/sync/mpsc/bounded.rs [hunk 1]
```diff
@@ -178,8 +184,18 @@ impl<T> Receiver<T> {
...
+
+        // If there is already an underflow, the permit released by recv is
+        // cancelled and the underflow is reduced accordingly.
+        if self.underflow > 0 {
+            self.chan.semaphore().reduce_permits(1);
+
+            self.underflow -= 1;
+        }
+
+        res
```
Comment (Darksonn) [inline tokio/src/sync/mpsc/bounded.rs:198]: This wont work. A sender may have consumed the permit before you get to the `reduce_permits` call, in which case the underflow is lost.

Comment (aym-v) [reply tokio/src/sync/mpsc/bounded.rs]: This did indeed open the door to race conditions, thank you. I added a parameter that allows the `recv` method to conditionally release a permit. This way, no sender can acquire a permit while the channel is overflowing. What do you think about this approach?

**Comment by Darksonn** [comment]
I'm sorry about being so slow to review this. Unfortunately, I still think the implementation is too complicated. The mpsc channel is already really complicated, probably too complicated, and adding even more complexity to the channel is something I want to avoid. 

I'm going to close this for now. If you want to continue working on the issue, please talk to us in the discord first so we can discuss whether the approach is sufficiently simple.

**Pull request closed without merging**